### PR TITLE
driver: adc: stm32: support dma for devices without dmamux

### LINKED ISSUE #65387: ADC via DMA on STM32:  DMA device is null and `dma_config()` leads to Instruction Address Violation
Hello Developers!

_this is my first issue to Zephyr, or any other open source project! if i should put this somewhere else or I just do it wrong -- tell me please._

I have a problem with ADC over DMA on STM32F429i-DISC1 board running samples/drivers/adc sample. I get a hardfault on dma_config(), executing an undefined instruction (api->config()). I think that something is not fully initialized through the DT.

I had the same issue on a STM32G071 on my custom board on DMAMUX and I couldn't solve it, so I tried a SoC (F429) without the MUX, but the problem prevails. My workaround was to use HAL/LL for DMA temporarily, but I would like to switch back to Zephyr implementation whenever possible. Any suggestions are much appreciated. 

I include a debug screenshot, call stack and zephyr terminal output, together with the DTS overlay and prj.conf entries.

Specs: 
Board: STM32f429i-Disc1
Build: `west build -p -b stm32f429i_disc1 samples/drivers/adc`  
Debug and flash tool: JLink OCD
Development environment: VSCode + cortex-debug
ZephyrOS version: 3.5.0
SDK: 0.16.3_windows_x86_64

### prj.conf entries:
```
CONFIG_ADC=y
CONFIG_DMA=y
CONFIG_ADC_STM32_DMA=y
CONFIG_ADC_ASYNC=y
```

### Debug screenshot (VSCode + cortex-debug via JLink)

> **Note:** `api=dev->api;` while `dev` is 0x00! Therefore `api` struct contains nonsense, and calling a function from a pointer inside it will lead to provblems

image

### Call Stack 
image

### Terminal screenshot
image

...and is my DTS overlay
```
 / {
    zephyr,user {
        io-channels = <&adc1 1>;//<&adc1 2 &adc1 1>;
    };
};

&adc1 {
    #address-cells = <1>;
    #size-cells = <0>;
    status = "okay";
    pinctrl-0 = <&gpio_a6_analog>;
    pinctrl-names = "default";
    st,adc-prescaler = <4>;
    st,adc-clock-source = <SYNC>;
    dmas = <&dma2 0 0 (STM32_DMA_PERIPH_TO_MEMORY | STM32_DMA_PERIPH_16BITS) 0x00>;
    dma-names="adcdma";
    channel@1 {
        reg = <1>;
        zephyr,gain = "ADC_GAIN_1";
        zephyr,reference = "ADC_REF_INTERNAL";
        zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
        zephyr,resolution = <12>;
    };

};

&pinctrl {
    gpio_a6_analog: gpio_a6_analog {
        pinmux = <STM32_PINMUX('A',6,ANALOG)>;
    };
    status = "okay";
};


&dma2 {
    status = "okay";
};
``` 

Hope putting this issue may help someone in the future

p.s. Used screenshots instead of text logs because I put the same problem on Discord forum before, and this is rather a copy of the post I made

### PR DESCRIPTION
**Pull request by heinwessels**
Previously the STM32 DMA driver was dependent on a very specific
name for the DMA in the DTS. This hidden requirement has caused
a bit of confusion. This commit changes the driver to instead
always use the first DMA listed in the ADC node's dma property.

Also, updates the dma code in the ADC driver to support the STM32L4 (and likely more other mcus)

Fixes: #65387

### CODE REVIEW

**Review by erwango — COMMENTED** [review]
Thanks for this work @heinwessels.

My understanding is that we're extending ADC DMA support to 'non dmamux' compatible series.

Would you mind adding at least target of this kind to tests/drivers/adc/adc_dma ?

**Comment by heinwessels** [comment]
> My understanding is that we're extending ADC DMA support to 'non dmamux' compatible series.

Yes, that's the goal.

> Would you mind adding at least target of this kind to tests/drivers/adc/adc_dma ?

I would like to, but it doesn't seem to work fully just yet. I found a possible problem which I mentioned on the Discord in the `#adc` channel. If I can verify that it does work I will add the test.

My plan was to merge this PR first as it works standalone IMO. It doesn't imply functionality without a dmamux. And once functionality is verified create a new PR with test (and maybe a sample) that has the title something explicit like "support non dmamux mcus".

Would you prefer I put everything into one PR?

**Comment by heinwessels** [comment]
@erwango I've added tests for the `nucleo_l476rg`. I can't run it though as I don't have this board. The person who brought up this issue initially said they might get time to run it on their hardware later this week.

File: tests/drivers/adc/adc_dma/boards/nucleo_l476rg.overlay [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2023 Hein Wessels, Nobleo Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&adc1 {
+	dmas = < &dma1 1 0 (STM32_DMA_MODE_NORMAL | STM32_DMA_PERIPH_TO_MEMORY |
+		STM32_DMA_MEM_INC |  STM32_DMA_MEM_16BITS | STM32_DMA_PERIPH_16BITS) >;
```
Comment (erwango) [inline tests/drivers/adc/adc_dma/boards/nucleo_l476rg.overlay:9]:
```suggestion
	dmas = < &dma1 1 0 (STM32_DMA_PERIPH_RX |  STM32_DMA_MEM_16BITS | STM32_DMA_PERIPH_16BITS) >;
```

Comment (heinwessels) [reply tests/drivers/adc/adc_dma/boards/nucleo_l476rg.overlay]: Done

File: tests/drivers/adc/adc_dma/boards/nucleo_l476rg.conf [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#
+# Copyright (c) 2023 Hein Wessels, Nobleo Technology
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_ADC_STM32_DMA=y
+CONFIG_ADC_ASYNC=y
```
Comment (erwango) [inline tests/drivers/adc/adc_dma/boards/nucleo_l476rg.conf:8]: This is not strictly related to this PR but I wonder if these 2 options should stay independent.
Can we have CONFIG_ADC_ASYNC=y and CONFIG_ADC_STM32_DMA=n ? or the other way round ?

**Review by erwango — DISMISSED** [review]
One pending question, but otherwise LGTM

File: tests/drivers/adc/adc_dma/boards/nucleo_l476rg.conf (hunk 2 above)
Comment (heinwessels) [reply tests/drivers/adc/adc_dma/boards/nucleo_l476rg.conf:8]: > Can we have CONFIG_ADC_ASYNC=y and CONFIG_ADC_STM32_DMA=n ?

I believe this is possible, and that these are independent. However, the reason `CONFIG_ADC_ASYNC=y` is because there's an async test part of this suite.

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -216,6 +216,35 @@ static bool init_irq = true;
...
+#elif defined(ADC_VER_V5_X)
+	LL_ADC_REG_SetDataTransferMode(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
+#else
+#error "Unsupported ADC version"
+#endif
+
+#else /* defined(CONFIG_SOC_SERIES_STM32H7X) */
+
+	/* Default mechanism for other MCUs */
+	LL_ADC_REG_SetDMATransfer(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
```
Comment (gautierg-st) [inline drivers/adc/adc_stm32.c:243]: A quick look through the LL shows that the macro `LL_ADC_REG_DMA_TRANSFER_LIMITED` does not exist for F1 and F37x.
For those, the choice is between `LL_ADC_REG_DMA_TRANSFER_NONE` or `LL_ADC_REG_DMA_TRANSFER_UNLIMITED`.
A `#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc)` should be added for these cases. What we do inside is up to you: unlimited seems to imply a circular buffer, which may be hard to add, so maybe we can just not support this case for now and return an error.

**Review by gautierg-st — CHANGES_REQUESTED** [review]
Need to add a special case for F1 compatibles, otherwise looks good

File: drivers/adc/adc_stm32.c (hunk 3 above)
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:243]: I would prefer to error as "not supported", as I do not have hardware or time to investigate all potential edge cases at the moment. And my idea was that my code as-is doesn't compile then it already implies "not supported", but I can make this explicit.

Comment (heinwessels) [reply drivers/adc/adc_stm32.c:247]: Added an explicit error

**Review by gautierg-st — APPROVED** [review]
Thanks

**Review by erwango — APPROVED** [review]
